src/services/catalog.service.ts

```diff
@@ -0,0 +1,198 @@
+            return alias === subCategory;
+          });
+          if (result !== -1) {
+            retSubCategory = cat_subCategory.id;
+            breakLoops = true;
+            return false;
+          }
+        }
+      });
+    });
+    return retSubCategory;
+  }
+}
+
+interface IServiceItem {
+  iconClass: string;
+  imageUrl: string;
+  name: string;
+  catsBySubCats: any;
+  description: string;
+  longDescription: string;
+  resource: any;
+}
+
+export class ServiceItem implements IServiceItem {
+  public iconClass: string;
+  public imageUrl: string;
+  public name: string;
+  public catsBySubCats: any;
+  public description: string;
+  public longDescription: string;
+  public resource: any;
+  private catalogSrv: CatalogService;
+
+  constructor (serviceClass: any, catalogSrv: CatalogService) {
+    this.resource = serviceClass;
+    this.catalogSrv = catalogSrv;
+    this.imageUrl = this.getImage();
+    this.name = this.getName();
+    this.description = this.getDescription();
+    this.longDescription = this.getLongDescription();
+    this.catsBySubCats = this.getCategoriesBySubCategories();
+  }
+
+  private getImage() {
+    if (this.resource.metadata) {
+      return this.resource.metadata.imageUrl;
+    }
+  }
+
+  private getName() {
+    if (this.resource.osbMetadata) {
+      return this.resource.osbMetadata.displayName ? this.resource.osbMetadata.displayName : this.resource.metadata.name;
+    } else {
+      return this.resource.metadata.name;
+    }
+  }
+
+  private getDescription() {
+    if (this.resource.osbMetadata) {
+      return this.resource.osbMetadata.description;
+    }
+  }
+
+  private getLongDescription() {
+    if (this.resource.osbMetadata) {
+      return this.resource.osbMetadata.longDescription;
+    }
+  }
+
+  private getCategoriesBySubCategories() {
+    let catsBySubCats = {};
+
+    if (this.resource.osbTags) {
+      _.each(this.resource.osbTags, (subCategory) => {
+        let subCat = this.catalogSrv.getSubCategoryFromAlias(subCategory);
+        let cat = this.catalogSrv.getCategoryFromSubCategory(subCat);
+        catsBySubCats[subCat] = cat;
+      });
+    }
+    return catsBySubCats;
+  }
+}
+
+export class ImageItem implements IServiceItem {
+  public iconClass: string;
+  public imageUrl: string;
+  public name: string;
+  public catsBySubCats: any;
+  public description: string;
+  public longDescription: string;
+  public resource: any;
+  public isBuilder: boolean;
+  private catalogSrv: CatalogService;
+
+  constructor (image: any, catalogSrv : CatalogService) {
+    this.resource = image;
+    this.catalogSrv = catalogSrv;
+    this.iconClass = this.getIcon();
+    this.name = this.getName();
+    this.description = this.getDescription();
+    this.longDescription = this.getLongDescription();
+    this.catsBySubCats = this.getCategoriesBySubCategories();
+  }
+
+  private getIcon() {
+    return 'font-icon ' + this.catalogSrv.$filter('imageStreamTagIconClass')(this.resource);
```

Comment:
repeating this comment that i put on the old file:
so i think i said before tagName was optional but I just realized that when you don't include it then it is assuming the tag is "latest" which we shouldn't really make the assumption that a latest tag exists :-\

so i think you need to actually grab the name of a valid tag off this imagestream, i think you can just do what you do above in annotations and just use the first tag you find in the list to get the name of a valid tag